internal/aws/cloudwatch/cwlog_client.go

```diff
@@ -0,0 +1,192 @@
+		response, err = client.svc.PutLogEvents(input)
+		if err != nil {
+			awsErr, ok := err.(awserr.Error)
+			if !ok {
+				client.logger.Error("Cannot cast PutLogEvents error into awserr.Error.", zap.Error(err))
+				return token, err
+			}
+			switch e := awsErr.(type) {
+			case *cloudwatchlogs.InvalidParameterException:
+				client.logger.Error("cwlog_client: Error occurs in PutLogEvents, will not retry the request", zap.Error(e), zap.String("LogGroupName", *input.LogGroupName), zap.String("LogStreamName", *input.LogStreamName))
+				return token, err
+			case *cloudwatchlogs.InvalidSequenceTokenException: //Resend log events with new sequence token when InvalidSequenceTokenException happens
+				client.logger.Warn("cwlog_client: Error occurs in PutLogEvents, will search the next token and retry the request", zap.Error(e))
+				token = e.ExpectedSequenceToken
+				continue
+			case *cloudwatchlogs.DataAlreadyAcceptedException: //Skip batch if DataAlreadyAcceptedException happens
+				client.logger.Warn("cwlog_client: Error occurs in PutLogEvents, drop this request and continue to the next request", zap.Error(e))
+				token = e.ExpectedSequenceToken
+				return token, err
+			case *cloudwatchlogs.OperationAbortedException: //Retry request if OperationAbortedException happens
+				client.logger.Warn("cwlog_client: Error occurs in PutLogEvents, will retry the request", zap.Error(e))
+				return token, err
+			case *cloudwatchlogs.ServiceUnavailableException: //Retry request if ServiceUnavailableException happens
+				client.logger.Warn("cwlog_client: Error occurs in PutLogEvents, will retry the request", zap.Error(e))
+				return token, err
+			case *cloudwatchlogs.ResourceNotFoundException:
+				tmpToken, tmpErr := client.CreateStream(input.LogGroupName, input.LogStreamName)
+				if tmpErr == nil {
+					if tmpToken == "" {
+						token = nil
+					}
+				}
+				continue
+			default:
+				// ThrottlingException is handled here because the type cloudwatch.ThrottlingException is not yet available in public SDK
+				// Drop request if ThrottlingException happens
+				if awsErr.Code() == errCodeThrottlingException {
+					client.logger.Warn("cwlog_client: Error occurs in PutLogEvents, will not retry the request", zap.Error(awsErr), zap.String("LogGroupName", *input.LogGroupName), zap.String("LogStreamName", *input.LogStreamName))
+					return token, err
+				}
+				client.logger.Error("cwlog_client: Error occurs in PutLogEvents", zap.Error(awsErr))
+				return token, err
+			}
+
+		}
+
+		//TODO: Should have metrics to provide visibility of these failures
+		if response != nil {
+			if response.RejectedLogEventsInfo != nil {
+				rejectedLogEventsInfo := response.RejectedLogEventsInfo
+				if rejectedLogEventsInfo.TooOldLogEventEndIndex != nil {
+					client.logger.Warn(fmt.Sprintf("%d log events for log group name are too old", *rejectedLogEventsInfo.TooOldLogEventEndIndex), zap.String("LogGroupName", *input.LogGroupName))
```

Comment:
Please don't use format strings inside a log statement, please reward it so the dynamic value is a field as part of the log message.